### PR TITLE
[SPARK-48901][SPARK-48916][SS][PYTHON] Introduce clusterBy DataStreamWriter API

### PR DESCRIPTION
**Pull request by chirag-s-db**
### What changes were proposed in this pull request?
Introduce a new clusterBy DataStreamWriter API in Scala, Spark Connect, and Pyspark.

### Why are the changes needed?
Provides another way for users to create clustered tables for streaming writes.

### Does this PR introduce _any_ user-facing change?
Yes, it adds a new clusterBy DataStreamWriter API in Scala, Spark Connect, and Pyspark to allow specifying the clustering columns when writing streaming DataFrames.

### How was this patch tested?
See new unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by chirag-s-db** [comment]
Note: this PR depends on https://github.com/apache/spark/pull/47301

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala [hunk 1]
```diff
@@ -303,7 +331,7 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
       val cmd = CreateTable(
         UnresolvedIdentifier(originalMultipartIdentifier),
         df.schema.asNullable.map(ColumnDefinition.fromV1Column(_, parser)),
-        partitioningColumns.getOrElse(Nil).asTransforms.toImmutableArraySeq,
+        partitioningOrClusteringTransform,
```
Comment (zedtang) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:334]: nit: just `partitioningTransform` should be fine, we don't have a concept of clusteringTransform.

**Review by zedtang — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala (hunk 1 above)
Comment (chirag-s-db) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:334]: Fixed.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 2]
```diff
@@ -471,6 +471,20 @@
     ],
     "sqlState" : "0A000"
   },
+  "CLUSTERING_COLUMNS_MISMATCH" : {
```
Comment (chirag-s-db) [inline common/utils/src/main/resources/error/error-conditions.json:474]: Changes copied from https://github.com/apache/spark/pull/47301

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 3]
```diff
@@ -209,10 +209,25 @@ object ClusterBySpec {
       normalizeClusterBySpec(schema, clusterBySpec, resolver).toJson
   }
 
+  /**
```
Comment (chirag-s-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:212]: Changes in this file copied from https://github.com/apache/spark/pull/47301

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 4]
```diff
@@ -1866,6 +1866,18 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
         "existingBucketString" -> existingBucketString))
   }
 
+  def mismatchedTableClusteringError(
```
Comment (chirag-s-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1869]: Changes in this file copied from https://github.com/apache/spark/pull/47301

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala [hunk 5]
```diff
@@ -194,6 +194,10 @@ abstract class InMemoryBaseTable(
           case (v, t) =>
             throw new IllegalArgumentException(s"Match: unsupported argument(s) type - ($v, $t)")
         }
+      case ClusterByTransform(columnNames) =>
```
Comment (chirag-s-db) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:197]: Changes in this file copied from https://github.com/apache/spark/pull/47301

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 6]
```diff
@@ -52,6 +52,11 @@ object DataSourceUtils extends PredicateHelper {
    */
   val PARTITION_OVERWRITE_MODE = "partitionOverwriteMode"
 
+  /**
```
Comment (chirag-s-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:55]: Changes in this file copied from https://github.com/apache/spark/pull/47301

**Comment by HeartSaVioR** [comment]
Please allow me reviewing after #47301 is merged and this PR is rebased. It seems to be around corner.

**Comment by chirag-s-db** [comment]
@HeartSaVioR https://github.com/apache/spark/pull/47301 has been merged, ready for review again!

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**